### PR TITLE
3732 costmap raw update msgs rebased (fixes github issues with PR #3948)

### PR DESCRIPTION
**Pull request by emSko**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #3732  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | gazebo simulation |

---

## Description of contribution in a few bullet points

* Adding new msg type **CostmapUpdate.msg**,
* Adding **CostmapUpdate** publisher to  **Costmap2DPublisher** class,
* Re-implementation of the **CostmapSubscriber** class to support the **CostmapUpdate** msg. 

## Description of documentation updates required from your changes

---

## Future work that may be required in bullet points

* I see a lot of redundancy in **Costmap2DPublisher** class while creating the msgs, adding methods that are responsible for populating msgs and return the unique ptr to msgs might improve the code readability

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Comment by emSko** [comment]
This is the continuation of work from #3948

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp [hunk 1]
```diff
@@ -181,13 +183,30 @@ void Costmap2DPublisher::prepareCostmap()
...
+  msg->x = x0_;
+  msg->y = y0_;
+  msg->size_x = xn_ - x0_;
+  msg->size_y = yn_ - y0_;
+  msg->data.resize(msg->size_x * msg->size_y);
+
+  std::uint32_t i = 0;
+  for (std::uint32_t y = y0_; y < yn_; y++) {
+    for (std::uint32_t x = x0_; x < xn_; x++) {
+      //todo copy from array instead of calling method
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_2d_publisher.cpp:201]: You have a few TODOs that should be resolved before turning this into a full non-draft PR.

**Review by SteveMacenski — COMMENTED** [review]
Otherwise, beyond tests, this looks good :-)

File: nav2_costmap_2d/src/costmap_subscriber.cpp [hunk 2]
```diff
@@ -14,84 +14,155 @@
...
 }
 
+// todo: change to const?
 std::shared_ptr<Costmap2D> CostmapSubscriber::getCostmap()
 {
   if (!costmap_received_) {
     throw std::runtime_error("Costmap is not available");
   }
-  toCostmap2D();
+  if (costmap_msg_ != nullptr) {
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_subscriber.cpp:67]:
```suggestion
  if (costmap_msg_) {
```

File: nav2_costmap_2d/src/costmap_subscriber.cpp [hunk 3]
```diff
@@ -14,84 +14,155 @@
...
+    costmap_->resizeMap(
+      costmap_msg_->metadata.size_x, costmap_msg_->metadata.size_y,
+      costmap_msg_->metadata.resolution,
+      costmap_msg_->metadata.origin.position.x,
+      costmap_msg_->metadata.origin.position.y);
   }
+
+  unsigned char * master_array = costmap_->getCharMap();
+  std::copy(costmap_msg_->data.begin(), costmap_msg_->data.end(), master_array);
+  costmap_msg_ = nullptr;
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_subscriber.cpp:141]: Use `costmap_msg_.reset()` will do the same thing and in the smart-pointer-y style

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp (hunk 1 above)
Comment (emSko) [reply nav2_costmap_2d/src/costmap_2d_publisher.cpp]: resolved with 4e42850620fa8b753ab0b8c7e0517cacf759a931

File: nav2_costmap_2d/src/costmap_subscriber.cpp (hunk 2 above)
Comment (emSko) [reply nav2_costmap_2d/src/costmap_subscriber.cpp]: resolved with 4e42850620fa8b753ab0b8c7e0517cacf759a931

File: nav2_costmap_2d/src/costmap_subscriber.cpp (hunk 3 above)
Comment (emSko) [reply nav2_costmap_2d/src/costmap_subscriber.cpp]: resolved with 4e42850620fa8b753ab0b8c7e0517cacf759a931

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp [hunk 4]
```diff
@@ -115,13 +117,20 @@ void Costmap2DPublisher::onNewSubscription(const ros::SingleSubscriberPublisher&
...
 
-// prepare grid_ message for publication.
-void Costmap2DPublisher::prepareGrid()
+
+void Costmap2DPublisher::updateGridParams()
 {
-  std::unique_lock<Costmap2D::mutex_t> lock(*(costmap_->getMutex()));
+  saved_origin_x_ = costmap_->getOriginX();
+  saved_origin_y_ = costmap_->getOriginY();
   grid_resolution = costmap_->getResolution();
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_2d_publisher.cpp:125]: Grid resolution/width/height should have trailing underscores as class member variables

**Comment by SteveMacenski** [comment]
Overall looks good / done to me! The costmap publisher seems to be missing coverage in several entire functions though: https://app.codecov.io/gh/ros-planning/navigation2/pull/3965/blob/nav2_costmap_2d/src/costmap_2d_publisher.cpp

**Comment by emSko** [comment]
@SteveMacenski  the lines that are missing coverage are not related to the functionality implemented in this issue (such as grid publishing and costmap service). I feel that lines were missing the coverage before my changes and the tests should be added in different PR

**Comment by SteveMacenski** [comment]
OK - still heed the underscore comment above

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp (hunk 4 above)
Comment (emSko) [reply nav2_costmap_2d/src/costmap_2d_publisher.cpp]: resolved with 9b92028155ceefbfa42881738e212f44109a26fe

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
@emSko last thing to merge this: Add new feature to the migration guide! https://navigation.ros.org/migration/Iron.html

Also, please do a run of the system and add some prints to make sure its working in production and not just sending the full costmap except on initialization (e.g. update callback is triggered at runtime mostly). Do you see any major computation changes in the runtime performance of the system? I suspect this change may actually be visible in metrics

**Comment by SteveMacenski** [comment]
Any metrics changes?

**Comment by emSko** [comment]
@SteveMacenski  I have run the simulation with the default map for Turtlebot3 and unfortunately, I could not observe any noticeable change in performance.
```
Full costmap not navigating. Average CPU time: 28.294999999999995, average memory 1.4000000000000004.

Full costmap navigating. Average CPU time: 49.1, average memory 1.3640000000000003.

Update not navigating. Average CPU time: 29.507, average memory 1.4000000000000004.

Update navigating. Average CPU time: 50.132, average memory 1.3770000000000002. 
```
Not navigating means that the robot was just standing still, navigating means that the robot was navigating to goals. These are the measurements for a composable node after collecting 100 samples for each case. 


Then with help from Alexey, I created a synthetic map of size 1000x1000 px, however, I was not able to collect the measurements because the simulation kept crashing on my laptop, but by just eyeballing the top command I would say that there still were not any observable changes.

**Comment by emSko** [comment]
I've checked the functionality with the ros2 topic echo command. It works as intended that is the full costmap is latched on the topic and only the updates are sent continuously. I also have implemented the helper node, as you suggested, that is converting the raw msgs into grid msgs to visualize it in rviz - it shows that everything is correct. 

From my analysis, I have found that it does not make sense to use the update msg for the local costmap. Due to the change of the origin while the robot is moving, the full costmap is sent and the updates are only sent when the robot is standing still. Also, the size of the full costmap is insignificantly larger than the update (3600 cells vs ~3300 cells). However, in the case of the global costmap, we are dealing with the decrease of the number of cells ~10 times (when running the simulation of the default turtlebot map/world, for larger maps the difference is larger).

**Comment by emSko** [comment]
I have nothing to add at the moment. As for me, the changes can be merged at the current state.

**Comment by emSko** [comment]
Thank you, Steven, for all the comments and suggestions.
I was thinking about working on the AI Integration Layers that are listed on the nav2 user survey. Could you provide me with more info about that on Slack, or can we discuss it during the working group meeting?

**Pull request merged**